philips_hue_mock.py

```diff
@@ -0,0 +1,52 @@
+# Copyright 2020 Google, LLC.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Module to provide callback functions for mock HTTP requests to the Philips Hue API."""
+
+import json
+
+
+def mock_hue_put_response(request, context):
+    """Callback for mocking a Philips Hue API response using the requests-mock library,
+    specifically for a put request.
+
+    See https://requests-mock.readthedocs.io/en/latest/response.html for usage details.
+
+    Args:
+        request: The requests.Request object that was provided.
+        context: An object containing the collected known data about this response
+        (headers, status_code, reason, cookies).
+
+    Returns:
+        The response text with confirmation of the arguments passed in.
+    """
+    try:
+        base_path_index = request.url.index('/lights/')
+    except ValueError:
+        context.status_code = 400
+        return 'invalid Philips Hue url'
+
+    try:
+        body_json = json.loads(request.body)
```

Comment:
I think that there are a few important things to somehow assert (feel free to check these however you feel best):

1. Is the URL correct?
2. Is the method correct? (get, put, etc)
3. Are the headers correct (I don't think we have any headers? So this can be omitted).
4. Is the body correct?

I could be missing this, but does this function ever check for the HTTP method and the request body?

Reply:
so requests_mock automatically matches the URL and method; by doing `requests_mock.put(f'http://{bridge_ip_address}/api/{username}/lights/1/state',
                      text=philips_hue_mock.mock_hue_put_response)`, a requests_mock exception will be thrown if the method or URL is incorrect. I suppose this probably isn't super useful though, since the mock callback won't even be called if the method or URL is incorrect, which means some of the code in philips_hue isn't actually being tested. I think I'll try to change the mock so that any kind of request goes through the callback, and that we can assert the criteria you listed above.

Reply:
For the request body so far I'm just checking that it's in the correct JSON encoded format, but I'll add extra checks to make sure that the body contains the correct keys as well.

Reply:
I updated the docstring for the mock function to be more clear on expected input.

Reply:
Okay, thanks for taking the time to explain this. Also, I'm not sure if this section of the docs is at all helpful:

https://requests-mock.readthedocs.io/en/latest/matching.html#path-matching